jl4/src/L4/TypeCheck.hs

```diff
@@ -1696,7 +1808,41 @@ unify (InfVar _ann1 _pre1 i1) t2
 unify t1 (InfVar _ann2 _pre2 i2)
   | i2 `elem` infVars t1 = pure False -- addError (OccursCheck t1 t2)
   | otherwise            = bind i2 t1
-unify _t1 _t2 = pure False -- addError (UnificationError t1 t2)
+unify t1 t2              = expandAndUnify t1 t2
+
+-- | Handles the cases where we've established we have no top-level
+-- inference variables.
+--
+expandAndUnify :: Type' Resolved -> Type' Resolved -> Check Bool
+expandAndUnify t1 t2 = do
+  mt1' <- tryExpand t1
+  case mt1' of
+    Just t1' -> unify t1' t2
+    Nothing  -> do
+      mt2' <- tryExpand t2
+      case mt2' of
+        Just t2' -> unify t1 t2'
+        Nothing  -> unifyBase t1 t2
+  where
+    tryExpand (TyApp _ann n ts) = tryExpandTypeSynonym n ts
+    tryExpand _                 = pure Nothing
```

Comment:
you could inline `tryExpandTypeSynonym` into `tryExpand` and you could use the `Alternative` instance of `MaybeT Check` to improve readability a bit.

Reply:
I can try. I'm not convinced it will be much more readable though.

Reply:
"Fixed" in d07e53bdfca651f35ad658e73e74ca11005cc788.